---
title: "1100-methylation"
author: "Mac Campbell"
date: "9/11/2022"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, warning = FALSE, message = FALSE)
```

## Using existing tools from Davis
https://ucdavis-bioinformatics-training.github.io/2020-Epigenetics_Workshop/WGBS/WGBS

First off creating a directory structure to mimic theirs. `/home/maccamp/Methylation`
with subdir '00-RawData`

We have two lanes of the same samples that neeed to be merged. Using something like 'PL1A01' /group/millermrgrp4/maccamp/smelt-bis/ group/millermrgrp4/maccamp/smelt-bis-2.    

In methyl-commands creating fastqs, not gzipping

```{sh, eval=FALSE}
cat ../meta/methyl-list.txt | while read line; do echo "gunzip -c /group/millermrgrp4/maccamp/smelt-bis/$line*.R1_001.fastq.gz > /home/maccamp/Methylation/00-RawData/$line-1.fastq" > methyl-commands-$line.sh; echo "gunzip -c /group/millermrgrp4/maccamp/smelt-bis-2/$line*.R1_001.fastq.gz >> /home/maccamp/Methylation/00-RawData/$line-1.fastq" >> methyl-commands-$line.sh; echo "gunzip -c /group/millermrgrp4/maccamp/smelt-bis/$line*.R2_001.fastq.gz > /home/maccamp/Methylation/00-RawData/$line-2.fastq" >> methyl-commands-$line.sh; echo "gunzip -c /group/millermrgrp4/maccamp/smelt-bis-2/$line*.R2_001.fastq.gz >> /home/maccamp/Methylation/00-RawData/$line-2.fastq" >> methyl-commands-$line.sh; done;
```       
       
` ls | grep sh | while read line; do echo "bash ./$line" >> parallel-commands; done;`    


Trying one, that is `(base) maccamp@farm:~/Methylation/00-RawData$ bash ~/dsm-omics/methyl-commands/methyl-commands-PL2H04.sh `

Batching:      
`srun -p high -t 1-10:00:00 --nodes=1 parallel -j 6 < parallel-commands > std.out 2>std.err &`   


## Reference
Using GCF_021917145.1_fHypTra1_genomic.fna
Also setting up chrom01 samtools faidx GCF_021917145.1_fHypTra1_genomic.fna NC_061060.1  >NC_061060.1.fna


