---
title: "1100-methylation"
author: "Mac Campbell"
date: "9/11/2022"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, warning = FALSE, message = FALSE)
```

## Using existing tools from Davis
https://ucdavis-bioinformatics-training.github.io/2020-Epigenetics_Workshop/WGBS/WGBS

First off creating a directory structure to mimic theirs. `/home/maccamp/Methylation`
with subdir '00-RawData`

We have two lanes of the same samples that neeed to be merged. Using something like 'PL1A01' /group/millermrgrp4/maccamp/smelt-bis/ group/millermrgrp4/maccamp/smelt-bis-2.    

In methyl-commands creating fastqs, not gzipping

```{sh, eval=FALSE}
cat ../meta/methyl-list.txt | while read line; do echo "gunzip -c /group/millermrgrp4/maccamp/smelt-bis/$line*R1_001.fastq.gz > /home/maccamp/Methylation/00-RawData/$line-1.fastq" > methyl-commands-$line.sh; echo "gunzip -c /group/millermrgrp4/maccamp/smelt-bis-2/$line*R1_001.fastq.gz >> /home/maccamp/Methylation/00-RawData/$line-1.fastq" >> methyl-commands-$line.sh; echo "gunzip -c /group/millermrgrp4/maccamp/smelt-bis/$line*R2_001.fastq.gz > /home/maccamp/Methylation/00-RawData/$line-2.fastq" >> methyl-commands-$line.sh; echo "gunzip -c /group/millermrgrp4/maccamp/smelt-bis-2/$line*R2_001.fastq.gz >> /home/maccamp/Methylation/00-RawData/$line-2.fastq" >> methyl-commands-$line.sh; done;
```       
       
` ls | grep sh | while read line; do echo "bash ./$line" >> parallel-commands; done;`    


Trying one, that is `(base) maccamp@farm:~/Methylation/00-RawData$ bash ~/dsm-omics/methyl-commands/methyl-commands-PL2H04.sh `

Batching:      
`srun -p high -t 1-10:00:00 --nodes=1 parallel -j 6 < parallel-commands > std.out 2>std.err &`   


## Reference
Using GCF_021917145.1_fHypTra1_genomic.fna
`(base) maccamp@farm:~/Methylation/scripts$ sbatch -J bismark-mac bismark_pre.slurm `

## Samples
Using meta/methyl-list.txt as allsamples.txt    

These were not sequenced twice:
(base) maccamp@farm:~/dsm-omics/methyl-commands$ cat std.err | cut -f 2 -d ":" | perl -pe 's/\/.*\///g' | perl -pe 's/\*.+$//g' | sort | uniq
     
     
Those samples copied over just fine though.     

## QC 
Initially running with 6 samples (have 133)

Running fastqc_pre.slurm

`(base) maccamp@farm:~/Methylation/scripts$ sbatch -J fastqc.$USER --array=1-6 fastqc_pre.slurm `
`(base) maccamp@farm:~/Methylation/scripts$ sbatch -p high -J mqc.${USER} multiqc_pre.slurm`
data looks good.     
Trimming      
`sbatch -p high -J tg.${USER} --array=1-6 trimgalore.slurm`     

Checking output post trimming.


jobid=$(sbatch -p high -J fqcp.${USER} --array=1-6 fastqc_post.slurm |cut -d' ' -f4 - )
sbatch -J mqcp.${USER} --dependency=afterok:${jobid} multiqc_post.slurm

## Alignment
sbatch -p high -J bm1.${USER} --array=1-133 bismark_part1.slurm # did not finish

      54947695_94       med bm1.macc  maccamp millermr  R       54:47  9-23:05:13     1 2   5G     c9-69
         54947695_15       med bm1.macc  maccamp millermr  R  5-04:01:44  4-19:58:16     1 2   5G     c9-74
         54947695_17       med bm1.macc  maccamp millermr  R  5-04:01:44  4-19:58:16     1 2   5G     c9-74
         
         
         
sbatch -p med -J bm1.${USER} --array=1-3 bismark_part1-finish.slurm 
tofinish.txt    
PL1H10    
PL1B03    
PL1B05    

## Array of bismark_part2

sbatch -p med -J bm2.${USER} --array=1-133 bismark_part2.slurm 

Looks to have run to completion.


Examine metrics:      
ls $HOME/Methylation/03-Bismark/*_PE_report.txt > input2.fofn
ls $HOME/Methylation/03-Bismark/*.deduplication_report.txt >> input2.fofn
ls $HOME/Methylation/04-Methylation/*.M-bias.txt >> input2.fofn
ls $HOME/Methylation/04-Methylation/*_splitting_report.txt >> input2.fofn
sbatch -J mqb.${USER} multiqc_bismark.slurm

Examine coverage with qualimap.
jobid=$(sbatch -J qms.${USER} --array=1-133 qualimap_single.slurm |cut -d' ' -f4 - )
sbatch -J qmm.${USER} --dependency=afterok:${jobid} qualimap_multi.slurm

# Getting counts in a different form
sbatch -J cnt.${USER} --array=1-133 gather_rawcounts.slurm

# DMRichR

https://www.benlaufer.com/DMRichR/articles/DMRichR.html#dmr-approach-and-interpretation

use the coverage2cytosine module in Bismark to generate the cytosine reports.     


