---
title: "1200-RNAseq"
author: "Mac Campbell"
date: "9/19/2022"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = FALSE, message = FALSE, warning=FALSE)
```

## RNA seq

Method? Machine?  Tagseq?

Looks to be single end data (expected)     
`ls ../00-RawData/ | perl -pe 's/.fastq.gz//g' > all-samples.txt`     

Existing workflows?     
https://github.com/z0on/tag-based_RNAseq
https://github.com/ben-laufer/Tag-seq/blob/master/tag-seq.sh    

__0__ Building STAR indices
module  star/2.7.10a 

Using gtf to annotate (https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/021/917/145/GCF_021917145.1_fHypTra1/GCF_021917145.1_fHypTra1_genomic.gtf.gz)

Read length = 85, sjdbOverhand 85-1 = 84

```{sh, eval=FALSE}
module star/2.7.10a 
srun -p high --nodes=1 --time=4:00:00 STAR --runThreadN 4 --runMode genomeGenerate --genomeDir /home/maccamp/RNAseq/reference/ --genomeFastaFiles /home/maccamp/RNAseq/reference/GCF_021917145.1_fHypTra1_genomic.fna --sjdbGTFfile /home/maccamp/RNAseq/reference/GCF_021917145.1_fHypTra1_genomic.gtf --sjdbOverhang 84 > index.out 2>index.err &
```

__1__ Basic assessment

Calling like this:     

```{sh eval=FALSE}
sbatch -p med -J fastqc.$USER --array=1-12 01-fastqc.slurm;
sbatch -J mqcp.${USER} 01-pre-multiqc.slurm
```


__2__ Trimming

We need to accomodate for quite a bit. trimmomatic or bbmap/bbduk.sh are appropriate. Requires some references for trimming such as Illumina RNA seq adapters, etc.    module avail indicates we have bbmap/38-72                  

Examining data for adapter signatures.    
>TruSeq Universal Adapter						
AATGATACGGCGACCACCGAGATCTACACTCTTTCCCTACACGACGCTCTTCCGATCT
>TruSeq Adapter, Index 1							
GATCGGAAGAGCACACGTCTGAACTCCAGTCACATCACGATCTCGTATGCCGTCTTCTGCTTG

```{sh, eval=FALSE}
gunzip -c AF104_UMI_S81_R1_001.fastq.gz | head -n 4000000 > temp.fastq
#HITS
#"GATCGGAAGAGCACACGTCTGAACTCCAGT", part of Truseq Adapter, index 1
#"AGATCGGAAGAGCACACGTCTGAACTCCAGTCA" Truseq forward
```

Adding these to trim
>TruSeq Universal Adapter						
AATGATACGGCGACCACCGAGATCTACACTCTTTCCCTACACGACGCTCTTCCGATCT
>TruSeq Adapter, Index 1							
GATCGGAAGAGCACACGTCTGAACTCCAGTCACATCACGATCTCGTATGCCGTCTTCTGCTTG
>TruSeq Adapter, Index 2							
GATCGGAAGAGCACACGTCTGAACTCCAGTCACCGATGTATCTCGTATGCCGTCTTCTGCTTG
>TruSeq Adapter, Index 3							
GATCGGAAGAGCACACGTCTGAACTCCAGTCACTTAGGCATCTCGTATGCCGTCTTCTGCTTG
>TruSeq Adapter, Index 4							
GATCGGAAGAGCACACGTCTGAACTCCAGTCACTGACCAATCTCGTATGCCGTCTTCTGCTTG
>TruSeq Adapter, Index 5							
GATCGGAAGAGCACACGTCTGAACTCCAGTCACACAGTGATCTCGTATGCCGTCTTCTGCTTG
>TruSeq Adapter, Index 6							
GATCGGAAGAGCACACGTCTGAACTCCAGTCACGCCAATATCTCGTATGCCGTCTTCTGCTTG
>TruSeq Adapter, Index 7							
GATCGGAAGAGCACACGTCTGAACTCCAGTCACCAGATCATCTCGTATGCCGTCTTCTGCTTG
>TruSeq Adapter, Index 8							
GATCGGAAGAGCACACGTCTGAACTCCAGTCACACTTGAATCTCGTATGCCGTCTTCTGCTTG
>TruSeq Adapter, Index 9							
GATCGGAAGAGCACACGTCTGAACTCCAGTCACGATCAGATCTCGTATGCCGTCTTCTGCTTG
>TruSeq Adapter, Index 10						
GATCGGAAGAGCACACGTCTGAACTCCAGTCACTAGCTTATCTCGTATGCCGTCTTCTGCTTG
>TruSeq Adapter, Index 11						
GATCGGAAGAGCACACGTCTGAACTCCAGTCACGGCTACATCTCGTATGCCGTCTTCTGCTTG
>TruSeq Adapter, Index 12						
GATCGGAAGAGCACACGTCTGAACTCCAGTCACCTTGTAATCTCGTATGCCGTCTTCTGCTTG
>Truseq forward read
AGATCGGAAGAGCACACGTCTGAACTCCAGTCA
>Truseq reverse read
AGATCGGAAGAGCGTCGTGTAGGGAAAGAGTGT

Trimming script 02-bbduk.slurm is this:

```{sh, eval=FALSE}
#!/bin/bash
#
#SBATCH --time=4-00
#SBATCH --mem=10G # Memory pool for all cores (see also --mem-per-cpu)
#SBATCH --output=slurmout/bbduk-%A-%a.out # File to which STDOUT will be written
#SBATCH --error=slurmout/bbduk-%A-%a.err # File to which STDERR will be written

start=`date +%s`
hostname

aklog
export baseP=/home/$USER/RNAseq/
export seqP=${baseP}/00-RawData
export outP=${baseP}/02-Trimmed
export cwd=${baseP}/scripts
export tmpP=$cwd/tmp


if [ ! -d "${outP}" ]; then
   mkdir ${outP}
fi

if [ ! -d "${tmpP}" ]; then
   mkdir ${tmpP}
fi


module load bbmap/38-72   

name=`head -n ${SLURM_ARRAY_TASK_ID} allsamples.txt | tail -1`


call="bbduk.sh -t ${SLURM_NTASKS} \
in=$seqP/${name}.fastq.gz \
out=$outP/${name}.fastq.gz  \
ref=adapters.fasta \
literal=AAAAAAAAAAAAAAAAAA \
k=13 \
ktrim=r \
useshortkmers=t \
mink=5 \
qtrim=r \
trimq=10 \
minlength=20 \
stats=$outP/${name}-stats.txt"

echo $call
eval $call

end=`date +%s`
runtime=$((end - start ))
echo $runtime
```

call:
`sbatch -p med -J bbduk$USER --array=1-12 02-bbduk.slurm`   

__3__ Post-trim fastqc

```{sh, eval=FALSE}
#!/bin/bash
#
#SBATCH --time=2-00
#SBATCH --mem=10G # Memory pool for all cores (see also --mem-per-cpu)
#SBATCH --output=slurmout/fqc-post-%A-%a.out # File to which STDOUT will be written
#SBATCH --error=slurmout/fqc-post-%A-%a.err # File to which STDERR will be written

start=`date +%s`
hostname

aklog
export baseP=/home/$USER/RNAseq/
export seqP=${baseP}/02-Trimmed
export outP=${baseP}/03-fastqc-post
export cwd=${baseP}/scripts
export tmpP=$cwd/tmp


if [ ! -d "${outP}" ]; then
   mkdir ${outP}
fi

if [ ! -d "${tmpP}" ]; then
   mkdir ${tmpP}
fi


module load fastqc/0.11.9

name=`head -n ${SLURM_ARRAY_TASK_ID} allsamples.txt | tail -1`


call="fastqc -t ${SLURM_NTASKS} --dir $tmpP --outdir $outP $seqP/${name}.fastq.gz"
echo $call
eval $call

#call="fastqc -t ${SLURM_NTASKS} --dir $tmpP --outdir $outP $seqP/${name}-2.fastq"
#echo $call
#eval $call


end=`date +%s`
runtime=$((end - start ))
echo $runtime
```

Note:
`(base) maccamp@farm:~/RNAseq/scripts$ gunzip -c ../00-RawData/AF104_UMI_S81_R1_001.fastq.gz  | wc `      
15308648 19135810 881001287     
`(base) maccamp@farm:~/RNAseq/scripts$ gunzip -c ../02-Trimmed/AF104_UMI_S81_R1_001.fastq.gz | wc`     
15283580 19104475 855775669     

`sbatch -p med -J fqp-$USER --array=1-12 03-post-trim-fastqc.slurm`   

Dropped dependency, see Methylation for better way to do this.    
`sbatch -J mqcp.${USER} 03-post-trim-multiqc.slurm`


```{sh, eval=FALSE}
#!/bin/bash
#
#SBATCH --time=1-00
#SBATCH --mem=5G # Memory pool for all cores (see also --mem-per-cpu)
#SBATCH --partition=bigmemm  # Partition to submit to
#SBATCH --account=millermrgrp
#SBATCH --output=slurmout/mqcp-%j.out # File to which STDOUT will be written
#SBATCH --error=slurmout/mqcp-%j.err # File to which STDERR will be written

start=`date +%s`
hostname

aklog
export baseP=$HOME/RNAseq
export seqP=${baseP}/03-fastqc-post
export outP=${baseP}/03-multiqc-post
export cwd=${baseP}/scripts
export tmpP=$cwd/tmp


if [ ! -d "${outP}" ]; then
   mkdir ${outP}
fi

if [ ! -d "${tmpP}" ]; then
   mkdir ${tmpP}
fi


#module load multiqc/1.9
#module load bio3
module load multiqc/bio3    

call="multiqc -o $outP/ -f $seqP/*_fastqc.zip"
echo $call
eval $call


end=`date +%s`
runtime=$((end - start ))
echo $runtime

```

__4__ Align



```{sh, eval=FALSE}
#!/bin/bash
#
#SBATCH --time=4-00
#SBATCH --mem=10G # Memory pool for all cores (see also --mem-per-cpu)
#SBATCH --output=slurmout/star-%A-%a.out # File to which STDOUT will be written
#SBATCH --error=slurmout/star-%A-%a.err # File to which STDERR will be written


start=`date +%s`
hostname
#aklog
export baseP=/home/$USER/RNAseq/
export seqP=${baseP}/02-Trimmed
export outP=${baseP}/04-Align
export cwd=${baseP}/scripts
export tmpP=$cwd/tmp

if [ ! -d "${outP}" ]; then
   mkdir ${outP}
fi

if [ ! -d "${tmpP}" ]; then
   mkdir ${tmpP}
fi

name=`head -n ${SLURM_ARRAY_TASK_ID} allsamples.txt | tail -1`

module load star/2.7.10a 

call="STAR \
--runThreadN 4 \
--genomeDir /home/maccamp/RNAseq/reference/ \
--readFilesIn $seqP/${name}.fastq.gz  \
--readFilesCommand zcat \
--outFilterType BySJout \
--outFilterMultimapNmax 20 \
--alignSJoverhangMin 8 \
--alignSJDBoverhangMin 1 \
--outFilterMismatchNmax 999 
--outFilterMismatchNoverLmax 0.1 \
--alignIntronMin 20 \
--alignIntronMax 1000000 \
--alignMatesGapMax 1000000 \
--outSAMattributes NH HI NM MD \
--outSAMtype BAM SortedByCoordinate \
--outFileNamePrefix $outP/${name}- \
--quantMode GeneCounts"

echo $call
eval $call
```

`sbatch -p med -J star-$USER --array=1-12 04-star-align.slurm `

Don't forget to index bams!!    
