---
title: "1300-sex-gwas"
output: html_document
date: "2023-05-09"
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, warning=FALSE, message=FALSE)
```

```{r}
library(tidyverse)
library(snpR)
```

Working in outputs/1300
We have bamlists/test80.bamlist
bamlists/test80.pheno

```{sh, eval=FALSE}
#Generate beagles
bash $HOME/dsm-omics/203-parallelize-pca.sh $HOME/dsm-omics/bamlists/test80.bamlist $HOME/dsm-omics/meta/1mbseqs.txt

#Generate cov matrix
#This way is pretty darn slow, I shouldn't do it this way again.
cat $HOME/dsm-omics/meta/1mbseqs.txt | while read line; do python $HOME/pcangsd/pcangsd.py -beagle $line-pca.beagle.gz -o $line -threads 10 > $line.stdout 2> $line.stderr; done;

#Make PCs, see 1000.1-makePCs.R 
#Execute in ~/dsm-omics/
Rscript 1300.1-makePCs.R

#Conduct chrom by chrom gwas, like 703.2, see 1000.1-doAsso2-withCov-perChr.sh. This is using the first three PCs as covariates
#Execute in ~/dsm-omics/outputs/1000/

bash $HOME/dsm-omics/1000.1-doAsso2-withCov-perChrom.sh $HOME/dsm-omics/bamlists/test91.bamlist  $HOME/dsm-omics/meta/1mbseqs.txt $HOME/dsm-omics/meta/test91.pheno 


```


Check lg 26
NC_061085.1   

```{r}
cov<-read_delim("outputs/1300/NC_061078.1.cov", col_names=FALSE, delim=" ") %>% as.matrix()
m80<-read_tsv("meta/meta.tsv")
bamlist<-read_tsv("bamlists/test80.bamlist", col_names = c("Path"))
phenos<-read_tsv("bamlists/test80.pheno", col_names=c("Pheno"))
bamlist$File<-gsub("/home/maccamp/dsm-omics/data/SOMM517/|/home/maccamp/dsm-omics/data/SOMM516/|/home/maccamp/dsm-omics/data/SOMM515/", "",bamlist$Path)

m80<-m80 %>% filter(SeqFile %in% bamlist$File)
m80<-bind_cols(m80, phenos)
m80$Pheno<-as.factor(m80$Pheno)
```

```{r}
m80r<-m80 %>% left_join(bamlist, by=c("SeqFile"="File")) %>% arrange(Pheno) 
m80r %>% select(Path) %>% write_tsv("bamlists/m80-ordered.bamlist", col_names = FALSE)
```


```{r}
#' @param samples character vector with the individuals IDs in the order in which
#' they were passed in the bamlist to angsd.
#' @param cov covariance matrix
covar2pcs <- function(samples, cov) {
  
  
  eig <- eigen(cov, symm = TRUE)
  PC <- as.data.frame(eig$vectors) %>%
    as_tibble() %>%
    setNames(sprintf("PC-%02d", 1:ncol(.)))
  
  samtib <- tibble(sample = samples)
  
  list(
    PCs = bind_cols(samtib, PC),
    eigevalues = eig$values
  )
}

```

```{r}
pca <- covar2pcs(m80$`Sample name`, cov)

pca_long <- pca$PCs %>%
  tidyr::gather(., key = "PC", "val", -sample)

# then expand a grid of the possible comparisons (ordered)
expg <- expand.grid(sample = pca$PCs$sample,
                    PCx = sprintf("PC-%02d", 1:6),
                    PCy = sprintf("PC-%02d", 1:6),
                    stringsAsFactors = FALSE) %>%
  tibble::as_tibble()

# then left join the pca results onto that
pca_pairs <- dplyr::left_join(expg, pca_long, by = c("sample", "PCx" = "PC")) %>%
  dplyr::rename(val_x = val) %>%
  dplyr::left_join(pca_long, by = c("sample", "PCy" = "PC")) %>%
  dplyr::rename(val_y = val)

pp_meta <- pca_pairs %>%   # just keep the first 6 PCs around
  left_join(., m80, by = c("sample" = "Sample name")) 

npc <- 3
pp_meta2 <- pp_meta %>%
  filter( (PCx %in% sprintf("PC-%02d", 1:npc)) & 
            (PCy %in% sprintf("PC-%02d", 1:npc)) )

eig <- eigen(cov, symm = TRUE)
var<-eig$values/sum(eig$values)
cumvar<-cumsum(eig$values)/sum(eig$values)


ggplot(pp_meta2) +
  geom_point(aes(x=val_x, y=val_y, shape=`library number`, color=Pheno)) +
  coord_fixed() +
  facet_grid (PCx ~ PCy)

```


```{r}
sub12<-pp_meta2 %>% filter( (PCx =="PC-01") & (PCy =="PC-02") )


pc12<-ggplot(sub12 %>% filter(Coverage < 30), aes(x = val_x, y = val_y, color = Coverage, shape=Pheno)) +
  geom_point(size = 2, alpha=0.75) +
  scale_fill_discrete(na.value = "white") + 
  theme_bw()+
  theme(panel.grid=element_blank())+
  xlab(paste("PC1", " ", round((100*var[1]),2), "%", sep = ""))+
  ylab(paste("PC2", " ", round((100*var[2]),2), "%", sep = ""))+
  ggtitle("PCs 1 & 2")+
  theme(plot.title = element_text(hjust = 0.5)) +
  scale_colour_viridis_c() 
pc12
```


## As function

```{r}
pcs<-function(chrom) {
pcseqs<-m80
cov<-read_delim(paste0("outputs/1300/",chrom,".cov"), col_names=FALSE, delim=" ") %>% as.matrix()

pca <- covar2pcs(pcseqs$`Sample name`, cov)

pca_long <- pca$PCs %>%
  tidyr::gather(., key = "PC", "val", -sample)

# then expand a grid of the possible comparisons (ordered)
expg <- expand.grid(sample = pca$PCs$sample,
                    PCx = sprintf("PC-%02d", 1:6),
                    PCy = sprintf("PC-%02d", 1:6),
                    stringsAsFactors = FALSE) %>%
  tibble::as_tibble()

# then left join the pca results onto that
pca_pairs <- dplyr::left_join(expg, pca_long, by = c("sample", "PCx" = "PC")) %>%
  dplyr::rename(val_x = val) %>%
  dplyr::left_join(pca_long, by = c("sample", "PCy" = "PC")) %>%
  dplyr::rename(val_y = val)

pp_meta <- pca_pairs %>%   # just keep the first 6 PCs around
  left_join(., pcseqs, by = c("sample" = "Sample name")) 

npc <- 3
pp_meta2 <- pp_meta %>%
  filter( (PCx %in% sprintf("PC-%02d", 1:npc)) & 
            (PCy %in% sprintf("PC-%02d", 1:npc)) )

eig <- eigen(cov, symm = TRUE)
var<-eig$values/sum(eig$values)
cumvar<-cumsum(eig$values)/sum(eig$values)


ggplot(pp_meta2) +
  geom_point(aes(x=val_x, y=val_y, shape=Pheno, color=Coverage)) +
  coord_fixed() +
  facet_grid (PCx ~ PCy) +
  theme_bw()

ggsave(paste0("outputs/1300/figs/",chrom,".jpeg"))

sub12<-pp_meta2 %>% filter( (PCx =="PC-01") & (PCy =="PC-02") )


pc12<-ggplot(sub12 %>% filter(Coverage < 30), aes(x = val_x, y = val_y, color = Coverage, shape=Pheno)) +
  geom_point(size = 2, alpha=0.75) +
  scale_fill_discrete(na.value = "white") + 
  theme_bw()+
  theme(panel.grid=element_blank())+
  xlab(paste("PC1", " ", round((100*var[1]),2), "%", sep = ""))+
  ylab(paste("PC2", " ", round((100*var[2]),2), "%", sep = ""))+
  ggtitle("PCs 1 & 2")+
  theme(plot.title = element_text(hjust = 0.5)) +
  scale_colour_viridis_c() 

pc12

ggsave(paste0("outputs/1300/figs/pc12-",chrom,".jpeg"))

}
```



Get chrom names:
```{r, eval=FALSE}
chroms<-read_tsv("meta/1mbseqs.txt", col_names = c("Chromosome"))
out<-lapply(chroms$Chromosome,pcs)
```

## The gwas in outputs/1300

`bash $HOME/dsm-omics/1300-do-asso.sh $HOME/dsm-omics/bamlists/test80.bamlist  $HOME/dsm-omics/meta/1mbseqs.txt $HOME/dsm-omics/bamlists/test80.pheno`

Read them all in    
```{r}
list<-list.files(path = "outputs/1300", pattern="lrt0.gz", full.names = TRUE)
lrts<-lapply(list, read_tsv) %>% bind_rows() %>% mutate(log10p = -log10(dchisq(LRT, df = 1))) %>%
  filter(log10p > 1)
```


```{r}
ggplot(lrts) + geom_point(aes(x=Position, y=log10p, color=log10p), alpha=0.75, cex=0.7)+
  geom_hline(yintercept = 6, lty=2, alpha=0.6) +
  scale_color_gradient(low="grey",high="skyblue") +
  theme_bw()+
  theme(axis.text.x= element_text(angle=45,hjust=1)) +
  theme(panel.grid = element_blank()) +
  ylab("-log10(p)") +
  ggtitle("Sex GWAS") +
  facet_wrap(.~Chromosome, scales = "free_x") +
  theme(plot.title = element_text(hjust=0.5))

ggsave("outputs/1300/prelim-sex-gwas-all-chroms.jpeg", width=14, height=14)
```

NC_061068.1-asso.lrt0.gz       

```{r}
lrt<-read_tsv("outputs/1300/NC_061078.1-asso.lrt0.gz") %>% bind_rows() %>% mutate(log10p = -log10(dchisq(LRT, df = 1))) %>%  filter(log10p > 1)

lrt %>% arrange(-log10p)
```

NC_061078.1	15197806	A	G	0.329659	30.328595


```{r}
ggplot(lrt) + geom_point(aes(x=Position, y=log10p, color=log10p), alpha=0.75, cex=0.7)+
  geom_hline(yintercept = 6, lty=2, alpha=0.6) +
  scale_color_gradient(low="grey",high="skyblue") +
  theme_bw()+
  theme(axis.text.x= element_text(angle=45,hjust=1)) +
  theme(panel.grid = element_blank()) +
  ylab("-log10(p)") +
  ggtitle("Sex GWAS") +
  facet_wrap(.~Chromosome, scales = "free_x") +
  theme(plot.title = element_text(hjust=0.5))
```

SNP calls?

```{sh, eval=FALSE}
srun -t 72:00:00 -p high --mem=32G --nodes=1 --ntasks-per-node=1 --cpus-per-task=12 $HOME/angsd/angsd -nthreads 24 \
-minInd 72 -bam bamlists/test80.bamlist \
-ref /home/maccamp/genomes/GCF021917145.1/GCF_021917145.1_fHypTra1_genomic.fna \
-r NC_061078.1:10000000- -out outputs/1300/snps-wgs-NC_061078.1 \
-minMaf 0.05 -minMapQ 20 -minQ 20 -GL 1 -doMajorMinor 1 -doMaf 1 -SNP_pval 1e-6 \
-doGeno 2 -doPost 1 -postCutoff 0.9 -doPlink 2  > outputs/1300/snps-wgs-NC_061078.1.out 2> outputs/1300/snps-wgs-NC_061078.1.err &


srun -t 72:00:00 -p high --mem=32G --nodes=1 --ntasks-per-node=1 --cpus-per-task=12 $HOME/angsd/angsd -nthreads 24 \
-minInd 60 -bam bamlists/m80-ordered.bamlist \
-ref /home/maccamp/genomes/GCF021917145.1/GCF_021917145.1_fHypTra1_genomic.fna \
-r NC_061078.1:15121030-15213438 -out outputs/1300/snps-wgs \
-minMaf 0.1 -minMapQ 20 -minQ 20 -GL 1 -doMajorMinor 1 -doMaf 1 -SNP_pval 1e-6 \
-doGeno 2 -doPost 1 -postCutoff 0.9 -doPlink 2  > outputs/1300/snps-wgs.out 2> outputs/1300/snps-wgs.err &

```

Split into male/females and split 

Generally, for each tag ID there is a cross number and unique ID is a six-digit number in which the first two digits is a generation number, the three in the middle is a cross number, and the last number is for male/ female (1 or 2). 
So let's split into males/females and see what the maximal SNP differences are -> at most 0.5

```{r}
m80 %>% select(`Sample name`) %>% write_tsv("meta/m80-sample-names.txt", col_names=FALSE)
m80 %>% arrange(Pheno) %>% select(`Sample name`) %>% write_tsv("meta/m80-sample-names-sorted.txt", col_names=FALSE)
females<-m80 %>% filter(Pheno==1)
females %>% select(`Sample name`) %>% write_tsv("meta/females.txt", col_names = FALSE)
males<-m80 %>% filter(Pheno==0)
males %>% select(`Sample name`) %>% write_tsv("meta/males.txt", col_names = FALSE)
```

12598 variants
```{sh, eval=FALSE}
plink --tped snps-wgs-NC_061078.1.tped --tfam snps-wgs-NC_061078.1.tfam  --out plink-binary --recode --allow-extra-chr --noweb
plink --ped plink-binary.ped --map plink-binary.map --recode vcf --allow-extra-chr -out plink
bgzip plink.vcf 
tabix plink.vcf.gz

module load bcftools
#renaming vcf
bcftools reheader --samples meta/m80-sample-names.txt -o outputs/1300/renamed.vcf.gz outputs/1300/plink.vcf.gz
bcftools view -S meta/m80-sample-names-sorted.txt outputs/1300/renamed.vcf.gz > outputs/1300/reordered.vcf

bcftools view -S meta/females.txt outputs/1300/renamed.vcf.gz > outputs/1300/females.vcf
bcftools view -S meta/males.txt outputs/1300/renamed.vcf.gz > outputs/1300/males.vcf

```


```{r}
snps<-read_vcf("outputs/1300/renamed.vcf.gz", sample.meta = m80 %>% select(`Sample name`, Pheno) %>% rename(Sample=`Sample name`))
```
```{r}
x <- calc_association(snps, response = "Pheno", method = "armitage")
```

```{r}
stats <- as.data.frame(get.snpR.stats(x = x, stats = "association")) 
stats%>% arrange(single.p_armitage_Pheno)
```
NC_061078.1	15121039	T	G	0.187580	23.661750 we saw before. So the calls aren't getting the highest associated snps.



# Male align

We need to align our data to the male assembly for comparision:
~/genomes/hypomesus-male-genbank

 maccamp@farm:~/dsm-omics/data/male-align$ cat ../../bamlists/test80.bamlist | perl -pe 's/_R1.sort.flt.bam//g'| while read line; do ln -s $line*fastq.gz . ; done;

bash ../../doAlign-zipped.sh to-align-2.txt  /home/maccamp/genomes/hypomesus-male-genbank/GCA_021870715.1_mHypTra1_genomic.fna

Hey, it worked! Path to data is `data/male-align/`    

Running a silly gwas.

```{r}
mAlign<-m80 %>% mutate(Path=paste0("data/male-align/",`Sample name`,".sort.flt.bam")) 
mAlign %>% select(Path) %>% write_tsv("bamlists/test80-male.bamlist", col_names=FALSE)
mAlign %>% select(Pheno) %>% write_tsv("bamlists/test80-male.pheno", col_names=FALSE)

```

in outputs/1300/male
`bash $HOME/dsm-omics/1300-do-asso.sh $HOME/dsm-omics/bamlists/test80-male.bamlist  $HOME/dsm-omics/meta/1mbseqs.txt $HOME/dsm-omics/bamlists/test80.pheno`