---
title: "200-basic-qc"
author: "Mac Campbell"
date: "1/11/2022"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, warning=FALSE, message=FALSE)
```

## Check out data.

Investigating a few samples with FastQC v0.11.9: SOMM515 hdi_A_10, sub-sampling:

```{sh, eval=FALSE}
(base) maccamp@farm:~/dsm-omics$ gunzip -c data/SOMM515/hdi_A_10_R1.fastq.gz | head -n 400000 > outputs/200/test1.fastq
(base) maccamp@farm:~/dsm-omics$ gunzip -c data/SOMM515/hdi_A_10_R2.fastq.gz | head -n 400000 > outputs/200/test2.fastq
(base) maccamp@farm:~/dsm-omics$ gunzip -c data/SOMM515/hdi_c_16_R1.fastq.gz | head -n 400000 > outputs/200/test3.fastq
(base) maccamp@farm:~/dsm-omics$ gunzip -c data/SOMM515/hdi_c_16_R2.fastq.gz | head -n 400000 > outputs/200/test4.fastq
(base) maccamp@farm:~/dsm-omics$ gunzip -c data/SOMM516/111392_R1.fastq.gz | head -n 400000 > outputs/200/test5.fastq
(base) maccamp@farm:~/dsm-omics$ gunzip -c data/SOMM516/111392_R2.fastq.gz | head -n 400000 > outputs/200/test6.fastq

```


Looks great!!!! Overall length nearly all 150, some decline in q scores at the end of read, but mean ~32. Not seeing a lot of adapter content.

Running 201-align-and-sort.sh ...

Then 202-compute-coverage.sh ...

Ensieh says:

Generally, for each tag ID there is a cross number and unique ID is a six-digit number in which the first two digits is a generation number, the three in the middle is a cross number, and the last number is for male/ female (1 or 2).
111392


## PCAs

Meta   
```{r}
df<-read_csv("sequenced-fish/samples-barcodes-01042022.csv")
```

Bring in coverage:

```{r}
files<-list.files(path="outputs/200", pattern="*.cov", full.names=TRUE)
list<-lapply(files, read_tsv, col_names=FALSE) 
names<-lapply(files, basename)
all<-mapply(c, list, names, SIMPLIFY = FALSE)
comb<-as_tibble(rbindlist(all, fill=T)) %>% rename(Coverage=X1, CovFile=V1)

comb$SeqFile<-gsub(".cov",".bam",comb$CovFile)
comb$`Sample name`<-gsub("_R1.sort.flt.bam","",comb$SeqFile)
comb<-select(comb, -CovFile)
```


```{r}
meta<-left_join(df, comb) %>% filter(Coverage !="NA") %>% mutate(DI=ifelse(str_detect(`Sample name`, '^h|^H'), "HighDI","LowDI")) 
nrow(meta)
mean(meta$Coverage)
```
```{r}
meta %>% group_by(`library number`) %>%
  summarize(AvgCoverage=mean(Coverage))
```

```{r}
ggplot(meta) +
  geom_histogram(aes(Coverage, fill=`library number`)) +
  theme_bw() +
  theme(panel.grid=element_blank()) +
  ylab("Count") +
  theme(axis.title = element_text(face="bold")) +
  scale_y_continuous(breaks=c(0,5,10,15,20,25))
```

Lanes very similar.     
```{r} 
pcseqs<- meta %>% filter(Coverage > 5)  %>% filter(str_detect(`Sample name`, '^h|^H|^l|^L')) %>%     
  mutate(String=paste0("/home/maccamp/dsm-omics/data/",`library number`,"/",SeqFile))

pcseqs%>% group_by(`library number`) %>% summarize(Count=n())
```
     
     
```{r}
write_tsv(pcseqs %>% select(String), col_names = FALSE, file = "bamlists/test125.bamlist")
```

in /outputs/203 runnign 203-parallelize-pca.sh    

```{sh, eval=FALSE}
gunzip -c lg01-pca.beagle.gz | head -n 1 > chroms.beagle 
gunzip -c lg*beagle.gz | grep -v "marker"  >> chroms.beagle
gzip chroms.beagle
srun -p bigmemm --mem=32G -t 03:00:00 python $HOME/pcangsd/pcangsd.py -beagle chroms.beagle.gz -o chroms -threads 10 > chroms.stdout 2>chroms.stderr &
```


```{r}
cov<-read_delim("outputs/203/chroms.cov", col_names=FALSE, delim=" ") %>% as.matrix()
```


```{r}
#' @param samples character vector with the individuals IDs in the order in which
#' they were passed in the bamlist to angsd.
#' @param cov covariance matrix
covar2pcs <- function(samples, cov) {
  
  
  eig <- eigen(cov, symm = TRUE)
  PC <- as.data.frame(eig$vectors) %>%
    as_tibble() %>%
    setNames(sprintf("PC-%02d", 1:ncol(.)))
  
  samtib <- tibble(sample = samples)
  
  list(
    PCs = bind_cols(samtib, PC),
    eigevalues = eig$values
  )
}

```

```{r}
pca <- covar2pcs(pcseqs$`Sample name`, cov)

pca_long <- pca$PCs %>%
  tidyr::gather(., key = "PC", "val", -sample)

# then expand a grid of the possible comparisons (ordered)
expg <- expand.grid(sample = pca$PCs$sample,
                    PCx = sprintf("PC-%02d", 1:6),
                    PCy = sprintf("PC-%02d", 1:6),
                    stringsAsFactors = FALSE) %>%
  tibble::as_tibble()

# then left join the pca results onto that
pca_pairs <- dplyr::left_join(expg, pca_long, by = c("sample", "PCx" = "PC")) %>%
  dplyr::rename(val_x = val) %>%
  dplyr::left_join(pca_long, by = c("sample", "PCy" = "PC")) %>%
  dplyr::rename(val_y = val)

pp_meta <- pca_pairs %>%   # just keep the first 6 PCs around
  left_join(., pcseqs, by = c("sample" = "Sample name")) 

npc <- 3
pp_meta2 <- pp_meta %>%
  filter( (PCx %in% sprintf("PC-%02d", 1:npc)) & 
            (PCy %in% sprintf("PC-%02d", 1:npc)) )

eig <- eigen(cov, symm = TRUE)
var<-eig$values/sum(eig$values)
cumvar<-cumsum(eig$values)/sum(eig$values)


ggplot(pp_meta2) +
  geom_point(aes(x=val_x, y=val_y, shape=`library number`, color=DI)) +
  coord_fixed() +
  facet_grid (PCx ~ PCy)

```


```{r}
sub12<-pp_meta2 %>% filter( (PCx =="PC-01") & (PCy =="PC-02") )


pc12<-ggplot(sub12, aes(x = val_x, y = val_y, color = Coverage, shape=DI)) +
  geom_point(size = 2, alpha=0.75) +
  scale_fill_discrete(na.value = "white") + 
  theme_bw()+
  theme(panel.grid=element_blank())+
  xlab(paste("PC1", " ", round((100*var[1]),2), "%", sep = ""))+
  ylab(paste("PC2", " ", round((100*var[2]),2), "%", sep = ""))+
  ggtitle("PCs 1 & 2")+
  theme(plot.title = element_text(hjust = 0.5)) +
  scale_colour_viridis_c() +
  facet_wrap(.~`library number`)

pc12
```

## Aligning the two references with nucmer

base) maccamp@farm:~/dsm-omics$ srun -p bigmemm --time 2-01:00:00 nucmer -p outputs/200/mummer ~/genomes/hypomesus-male/mHypTra1.fasta ~/genomes/hypomesus-20210204/Hyp_tra_F_20210204.fa > outputs/200/mummer.stdout 2> outputs/200/mummer.stderr

To view a summary of all the alignments produced by NUCmer, we need to run the nucmer.delta file through the show-coords utility.

srun -p high --time 0-01:00:00 show-coords -r -c -l outputs/200/mummer.delta > outputs/200/mummer.coords     
srun -p high --time 0-01:00:00 show-tiling outputs/200/mummer.delta > outputs/200/mummer.tiling      

Plotting:
mummerplot -postscript -p outputs/200/mummerplot outputs/200/mummer.delta 
mummerplot -postscript -p outputs/200/mummertiling outputs/200/mummer.tiling

#LASTZ 

srun -p bigmemm -t 1-12:00:00 --mem=16G lastz outputs/200/male27.fasa ~/genomes/hypomesus-20210204/Hyp_tra_F_20210204.fa --gapped --ambiguous=iupac --gfextend --identity=75.0..100.0 --format=general > outputs/200/male27.lastz 2> outputs/200/male27.stderr &      

srun -p bigmemm -t 1-12:00:00 --mem=16G lastz outputs/200/male28.fasa ~/genomes/hypomesus-20210204/Hyp_tra_F_20210204.fa --gapped --ambiguous=iupac --gfextend --identity=75.0..100.0 --format=general> outputs/200/male28.lastz 2> outputs/200/male28.stderr &      


