---
title: "200-basic-qc"
author: "Mac Campbell"
date: "1/11/2022"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, warning=FALSE, message=FALSE)
```

## Check out data.

Investigating a few samples with FastQC v0.11.9: SOMM515 hdi_A_10, sub-sampling:

```{sh, eval=FALSE}
(base) maccamp@farm:~/dsm-omics$ gunzip -c data/SOMM515/hdi_A_10_R1.fastq.gz | head -n 400000 > outputs/200/test1.fastq
(base) maccamp@farm:~/dsm-omics$ gunzip -c data/SOMM515/hdi_A_10_R2.fastq.gz | head -n 400000 > outputs/200/test2.fastq
(base) maccamp@farm:~/dsm-omics$ gunzip -c data/SOMM515/hdi_c_16_R1.fastq.gz | head -n 400000 > outputs/200/test3.fastq
(base) maccamp@farm:~/dsm-omics$ gunzip -c data/SOMM515/hdi_c_16_R2.fastq.gz | head -n 400000 > outputs/200/test4.fastq
(base) maccamp@farm:~/dsm-omics$ gunzip -c data/SOMM516/111392_R1.fastq.gz | head -n 400000 > outputs/200/test5.fastq
(base) maccamp@farm:~/dsm-omics$ gunzip -c data/SOMM516/111392_R2.fastq.gz | head -n 400000 > outputs/200/test6.fastq

```


Looks great!!!! Overall length nearly all 150, some decline in q scores at the end of read, but mean ~32. Not seeing a lot of adapter content.

Running 201-align-and-sort.sh ...

Then 202-compute-coverage.sh ...

Ensieh says:

Generally, for each tag ID there is a cross number and unique ID is a six-digit number in which the first two digits is a generation number, the three in the middle is a cross number, and the last number is for male/ female (1 or 2).
111392


## Aligning the two references with nucmer

base) maccamp@farm:~/dsm-omics$ srun -p bigmemm --time 2-01:00:00 nucmer -p outputs/200/mummer ~/genomes/hypomesus-male/mHypTra1.fasta ~/genomes/hypomesus-20210204/Hyp_tra_F_20210204.fa > outputs/200/mummer.stdout 2> outputs/200/mummer.stderr

To view a summary of all the alignments produced by NUCmer, we need to run the nucmer.delta file through the show-coords utility.

srun -p high --time 0-01:00:00 show-coords -r -c -l outputs/200/mummer.delta > outputs/200/mummer.coords     
srun -p high --time 0-01:00:00 show-tiling outputs/200/mummer.delta > outputs/200/mummer.tiling      

Plotting:
mummerplot -postscript -p outputs/200/mummerplot outputs/200/mummer.delta 
mummerplot -postscript -p outputs/200/mummertiling outputs/200/mummer.tiling

#LASTZ 

srun -p bigmemm -t 1-12:00:00 --mem=16G lastz outputs/200/male27.fasa ~/genomes/hypomesus-20210204/Hyp_tra_F_20210204.fa --gapped --ambiguous=iupac --gfextend --identity=75.0..100.0 --format=general > outputs/200/male27.lastz 2> outputs/200/male27.stderr &      

srun -p bigmemm -t 1-12:00:00 --mem=16G lastz outputs/200/male28.fasa ~/genomes/hypomesus-20210204/Hyp_tra_F_20210204.fa --gapped --ambiguous=iupac --gfextend --identity=75.0..100.0 --format=general> outputs/200/male28.lastz 2> outputs/200/male28.stderr &      


