---
title: "300-gwas"
author: "Mac Campbell"
date: "1/18/2022"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, warning=FALSE, message=FALSE)
```

```{r}
library(tidyverse)
library(data.table)
```

## Getting a list of seqs
`cat  genomes/hypomesus-20210204/Hyp_tra_F_20210204.fa.fai | awk '$2 > 100000' | cut -f 1 > genomes/hypomesus-20210204/seqs100k.txt`



## Get some idea of what we got

```{r}
df<-read_csv("sequenced-fish/samples-barcodes-01042022.csv") %>% filter(str_detect(`Sample name`, '^\\d+')) %>%
  separate(col = `Sample name`, into=c("X1","X2","X3","X4","X5","X6","X7"), sep="", remove=FALSE ) %>%
  mutate(Sex = ifelse(X7==2, "female","male")) %>%
  mutate(Pheno=ifelse(Sex=="female", 1, 0))
```

first two digits is a generation number, the three in the middle is a cross number, and the last number is for male/ female (1 or 2).

Bring in coverage:


Meta   
```{r}
df<-read_csv("sequenced-fish/samples-barcodes-01042022.csv")
```

Bring in coverage:

```{r}
files<-list.files(path="outputs/600", pattern="*.cov", full.names=TRUE)
list<-lapply(files, read_tsv, col_names=FALSE) 
names<-lapply(files, basename)
all<-mapply(c, list, names, SIMPLIFY = FALSE)
comb<-as_tibble(rbindlist(all, fill=T)) %>% rename(Coverage=X1, CovFile=V1)

comb$SeqFile<-gsub(".cov",".bam",comb$CovFile)
comb$`Sample name`<-gsub("_R1.sort.flt.bam","",comb$SeqFile)
comb<-select(comb, -CovFile)
```


```{r}
meta<-left_join(df, comb) %>% filter(Coverage !="NA") %>% mutate(DI=ifelse(str_detect(`Sample name`, '^h|^H'), "HighDI","LowDI")) 
nrow(meta)
mean(meta$Coverage)
write_tsv(meta, "meta/dsm-meta.tsv")
```    

```{r}
meta %>% group_by(`library number`) %>%
  summarize(AvgCoverage=mean(Coverage))
```

```{r}
ggplot(meta) +
  geom_histogram(aes(Coverage, fill=`library number`)) +
  theme_bw() +
  theme(panel.grid=element_blank()) +
  ylab("Count") +
  theme(axis.title = element_text(face="bold")) +
  scale_y_continuous(breaks=c(0,5,10,15,20,25))
```

```{r}
files<-list.files(path="outputs/300", pattern="*.cov", full.names=TRUE)
list<-lapply(files, read_tsv, col_names=FALSE) 
names<-lapply(files, basename)
all<-mapply(c, list, names, SIMPLIFY = FALSE)
comb<-as_tibble(rbindlist(all, fill=T)) %>% rename(Coverage=X1, CovFile=V1)

comb$SeqFile<-gsub(".cov",".bam",comb$CovFile)
comb$`Sample name`<-gsub("_R1.sort.flt.bam","",comb$SeqFile)
comb<-select(comb, -CovFile)
```

Merge   
```{r}
df<-left_join(df,comb) %>% filter(Coverage !="NA")
nrow(df)
mean(df$Coverage)
```

```{r}
ggplot(df) +
  geom_histogram(aes(Coverage, fill=Sex)) +
  theme_bw() +
  theme(panel.grid=element_blank()) +
  ylab("Count") +
  theme(axis.title = element_text(face="bold")) +
  scale_y_continuous(breaks=c(0,5,10,15,20,25))
```


Creating a tester bamlit

```{r}
tester<-df %>% filter(Coverage > 10) %>% filter(Coverage < 30) %>% mutate(String=paste0("/home/maccamp/dsm-omics/data/",`library number`,"/",SeqFile))
tester %>% group_by(Sex) %>% summarize(Count=n())
```

```{r}
mean(tester$Coverage)
```
    
```{r}
write_tsv(tester %>% select(String), col_names = FALSE, file = "bamlists/test92.bamlist")
write_tsv(tester %>% select(Pheno), col_names=FALSE, file="bamlists/test92.pheno")
```

Running PCA in outputs/303


## Results of GWAS

```{r}
d01<-read_tsv("outputs/300/lg01-asso.lrt0.gz")
d02<-read_tsv("outputs/300/lg02-asso.lrt0.gz")
d03<-read_tsv("outputs/300/lg03-asso.lrt0.gz")
d19<-read_tsv("outputs/300/lg19-asso.lrt0.gz") 

d19$Chromosome<-gsub("lg19","Mystery Chromosome", d19$Chromosome)

data<-bind_rows(d01,d02,d03,d19) %>% mutate(log10p = -log10(dchisq(LRT, df = 1))) %>%
  filter(log10p > 1)
```

```{r}
ggplot(data) + geom_point(aes(x=Position, y=log10p, color=log10p), alpha=0.75, cex=0.7)+
  geom_hline(yintercept = 6, lty=2, alpha=0.6) +
  scale_color_gradient(low="grey",high="skyblue") +
  theme_bw()+
  theme(axis.text.x= element_text(angle=45,hjust=1)) +
  theme(panel.grid = element_blank()) +
  ylab("-log10(p)") +
  ggtitle("Sex GWAS") +
  facet_wrap(.~Chromosome, scales = "free_x") +
  theme(plot.title = element_text(hjust=0.5))

ggsave("outputs/300/prelim-sex-gwas.pdf")
```

Read them all in    
```{r}
list<-list.files(path = "outputs/300", pattern="lrt0.gz", full.names = TRUE)
lrts<-lapply(list, read_tsv) %>% bind_rows() %>% mutate(log10p = -log10(dchisq(LRT, df = 1))) %>%
  filter(log10p > 1)
```


```{r}
ggplot(lrts) + geom_point(aes(x=Position, y=log10p, color=log10p), alpha=0.75, cex=0.7)+
  geom_hline(yintercept = 6, lty=2, alpha=0.6) +
  scale_color_gradient(low="grey",high="skyblue") +
  theme_bw()+
  theme(axis.text.x= element_text(angle=45,hjust=1)) +
  theme(panel.grid = element_blank()) +
  ylab("-log10(p)") +
  ggtitle("Sex GWAS") +
  facet_wrap(.~Chromosome, scales = "free_x") +
  theme(plot.title = element_text(hjust=0.5))

ggsave("outputs/300/prelim-sex-gwas-all-chroms.jpeg", width=14, height=14)
```

## Comparing to RAD data    

(base) maccamp@farm:~/delta-smelt$ srun -p bigmemm -t 02:00:00 --nodes=1 $HOME/angsd/angsd -P 6 -bam $HOME/delta-smelt/bamlists/2012.bamlist -ybin outputs/temp/sex-phenos.txt -minMapQ 20 -minQ 20 -minInd 91 -doAsso 1 -GL 1 -doMajorMinor 1 -doMaf 1 -SNP_pval 1e-6 -r lg19 -out outputs/temp/rad-sex
