---
title: "600-new-genome-align"
author: "Mac Campbell"
date: "3/15/2022"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE,warning=FALSE,message=FALSE)
```


```{r}
library(tidyverse)
library(viridis)
```

Old data mv *bam *bai *cov *sh *out /group/millermrgrp4/maccamp/smelt-wgs/original-alignments/     
Dropping unmatched data     

## Aligning to the new genome
in each subdir

bash ../../201-align-and-sort.sh sequences.txt $HOME/genomes/GCF021917145.1/GCF_021917145.1_fHypTra1_genomic.fna  


## Computing coverage
     
See 202-compute-coverage.sh, doing by chrom with a bed file for the entire assembly.    

`cat GCF_021917145.1_fHypTra1_genomic.fna.fai  | cut -f 1,2 | awk '{print $1, 0, $2}' > GCF_021917145.1_fHypTra1_genomic.fna.bed`

../../202-compute-coverage.sh list.txt $HOME/genomes/GCF021917145.1/GCF_021917145.1_fHypTra1_genomic.fna.bed
 

Meta   
```{r}
df<-read_csv("sequenced-fish/samples-barcodes-01042022.csv")
```

Bring in coverage:

```{r}
files<-list.files(path="outputs/600", pattern="*.cov", full.names=TRUE)
list<-lapply(files, read_tsv, col_names=FALSE) 
names<-lapply(files, basename)
all<-mapply(c, list, names, SIMPLIFY = FALSE)
comb<-as_tibble(rbindlist(all, fill=T)) %>% rename(Coverage=X1, CovFile=V1)

comb$SeqFile<-gsub(".cov",".bam",comb$CovFile)
comb$`Sample name`<-gsub("_R1.sort.flt.bam","",comb$SeqFile)
comb<-select(comb, -CovFile)
```


```{r}
meta<-left_join(df, comb) %>% filter(Coverage !="NA") %>% mutate(DI=ifelse(str_detect(`Sample name`, '^h|^H'), "HighDI",
                                                                           ifelse(str_detect(`Sample name`, '^l|^L'), "LowDI", "Epi")))
nrow(meta)
mean(meta$Coverage)
write_tsv(meta, "meta/meta-ncbi.tsv")
```

Removing epigenetics fish for summarizing crosses   
```{r}
sum<-meta %>% filter(DI != "Epi") %>% separate(`Sample name`, into=c("CrossType","Tank","Number"), remove=FALSE, sep="_")
sum$Tank<-toupper(sum$Tank)
sum$CrossType<-toupper(sum$CrossType)

sum %>% select(`Sample name`, CrossType, Tank, Number) %>% group_by(CrossType, Tank) %>% summarize(Count=n())

sum %>% select(`Sample name`, CrossType, Tank, Number) %>% group_by(CrossType) %>% summarize(Count=n())

write_csv(sum, "meta/meta-excl-epi-ncbi.csv")
```



```{r}
meta %>% group_by(`library number`) %>%
  summarize(AvgCoverage=mean(Coverage))
```

```{r}
ggplot(meta) +
  geom_histogram(aes(Coverage, fill=`library number`)) +
  theme_bw() +
  theme(panel.grid=element_blank()) +
  ylab("Count") +
  theme(axis.title = element_text(face="bold")) +
  scale_y_continuous(breaks=c(0,5,10,15,20,25))
```

Similar coverage overall!!

Coverage excluding epi samples
```{r}
ggplot(sum) +
  geom_histogram(aes(Coverage, fill=`library number`)) +
  theme_bw() +
  theme(panel.grid=element_blank()) +
  ylab("Count") +
  theme(axis.title = element_text(face="bold")) +
  scale_y_continuous(breaks=c(0,5,10,15,20,25))

ggsave("outputs/600/average-coverage.jpeg")
```
     
```{r}
sum %>% group_by(`library number`) %>%
  summarize(AvgCoverage=mean(Coverage))
```


## Creating a  test dataset


```{r}
pcseqs<- sum %>% filter(Coverage > mean(sum$Coverage) - sd(sum$Coverage)) %>%
  filter(Coverage < mean(sum$Coverage) + sd(sum$Coverage)) %>%
  mutate(String=paste0("/home/maccamp/dsm-omics/data/",`library number`,"/",SeqFile))

pcseqs <- pcseqs %>% separate(`Sample name`, into=c("Type","Letter","Number"), remove=FALSE, sep="_" ) %>%
  mutate(Individual=1:n())

pcseqs%>% group_by(`library number`) %>% summarize(Count=n())
pcseqs%>% group_by(CrossType) %>% summarize(Count=n())

```
     
```{r}
nrow(pcseqs)
mean(pcseqs$Coverage)
sd(pcseqs$Coverage)
```

```{r}
write_tsv(pcseqs %>% select(String), col_names = FALSE, file = "bamlists/test96.bamlist")
```
     
     
Running  203-parallelize-pca.sh in outputs/601

`cat ~/genomes/GCF021917145.1/GCF_021917145.1_fHypTra1_genomic.fna.fai  | awk '$2 > 1000000 {print ;}' | cut -f 1 > 1mbseqs.txt`     

`bash $HOME/dsm-omics/203-parallelize-pca.sh $HOME/dsm-omics/bamlists/test96.bamlist $HOME/dsm-omics/meta/1mbseqs.txt`     

Set up beagle file:

```{sh, eval=FALSE}
gunzip -c NC_061060.1-pca.beagle.gz | head -n 1 > chroms.beagle 
gunzip -c NC*beagle.gz | grep -v "marker"  >> chroms.beagle
gunzip -c NW*beagle.gz | grep -v "marker" >> chroms.beagle
gzip chroms.beagle

srun -p bigmemm --mem=32G -t 03:00:00 python $HOME/pcangsd/pcangsd.py -beagle chroms.beagle.gz -o chroms -threads 10 > chroms.stdout 2>chroms.stderr &
```


PCAngsd 0.982
Using 10 thread(s)

Parsing Beagle file
Read 96 samples and 1616109 sites

Estimating population allele frequencies
EM (MAF) converged at iteration: 11

Number of sites after MAF filtering (0.05): 1616091


```{sh, eval=FALSE}
srun -p bigmemm --mem=32G -t 03:00:00 python $HOME/pcangsd/pcangsd.py -beagle outputs/601/chroms.beagle.gz -kinship -o outputs/601/test96.kinship -threads 10 > outputs/601/pca.stdout 2> outputs/601/pca.stderr &

srun -p bigmemm --mem=32G -t 03:00:00 python $HOME/pcangsd/pcangsd.py -beagle outputs/601/chroms.beagle.gz -relate outputs/601/test96.kinship.kinship.npy -o outputs/601/test96-kinremoved -threads 10 > outputs/601/pca-kin.stdout 2> outputs/601/pca-kin.stderr &
```


Downloading cov matrix.



```{r}
cov<-read_delim("outputs/601/chroms.cov", col_names=FALSE, delim=" ") %>% as.matrix()
```


```{r}
#' @param samples character vector with the individuals IDs in the order in which
#' they were passed in the bamlist to angsd.
#' @param cov covariance matrix
covar2pcs <- function(samples, cov) {
  
  
  eig <- eigen(cov, symm = TRUE)
  PC <- as.data.frame(eig$vectors) %>%
    as_tibble() %>%
    setNames(sprintf("PC-%02d", 1:ncol(.)))
  
  samtib <- tibble(sample = samples)
  
  list(
    PCs = bind_cols(samtib, PC),
    eigevalues = eig$values
  )
}

```

```{r}
pca <- covar2pcs(pcseqs$`Sample name`, cov)

pca_long <- pca$PCs %>%
  tidyr::gather(., key = "PC", "val", -sample)

# then expand a grid of the possible comparisons (ordered)
expg <- expand.grid(sample = pca$PCs$sample,
                    PCx = sprintf("PC-%02d", 1:6),
                    PCy = sprintf("PC-%02d", 1:6),
                    stringsAsFactors = FALSE) %>%
  tibble::as_tibble()

# then left join the pca results onto that
pca_pairs <- dplyr::left_join(expg, pca_long, by = c("sample", "PCx" = "PC")) %>%
  dplyr::rename(val_x = val) %>%
  dplyr::left_join(pca_long, by = c("sample", "PCy" = "PC")) %>%
  dplyr::rename(val_y = val)

pp_meta <- pca_pairs %>%   # just keep the first 6 PCs around
  left_join(., pcseqs, by = c("sample" = "Sample name")) 

npc <- 3
pp_meta2 <- pp_meta %>%
  filter( (PCx %in% sprintf("PC-%02d", 1:npc)) & 
            (PCy %in% sprintf("PC-%02d", 1:npc)) )

eig <- eigen(cov, symm = TRUE)
var<-eig$values/sum(eig$values)
cumvar<-cumsum(eig$values)/sum(eig$values)


ggplot(pp_meta2) +
  geom_point(aes(x=val_x, y=val_y, shape=`library number`, color=DI)) +
  coord_fixed() +
  facet_grid (PCx ~ PCy)

```


```{r}
sub12<-pp_meta2 %>% filter( (PCx =="PC-01") & (PCy =="PC-02") )


pc12<-ggplot(sub12 %>% filter(Coverage < 30), aes(x = val_x, y = val_y, color = Coverage, shape=DI)) +
  geom_point(size = 2, alpha=0.75) +
  scale_fill_discrete(na.value = "white") + 
  theme_bw()+
  theme(panel.grid=element_blank())+
  xlab(paste("PC1", " ", round((100*var[1]),2), "%", sep = ""))+
  ylab(paste("PC2", " ", round((100*var[2]),2), "%", sep = ""))+
  ggtitle("PCs 1 & 2")+
  theme(plot.title = element_text(hjust = 0.5)) +
  scale_colour_viridis_c() +
  facet_wrap(.~`library number`)

pc12
```

```{r}
pc12<-ggplot(sub12, aes(x = val_x, y = val_y, color = DI, shape=`library number`)) +
  geom_point(size = 2, alpha=0.75) +
  scale_fill_discrete(na.value = "white") + 
  theme_bw()+
  theme(panel.grid=element_blank())+
  xlab(paste("PC1", " ", round((100*var[1]),2), "%", sep = ""))+
  ylab(paste("PC2", " ", round((100*var[2]),2), "%", sep = ""))+
  ggtitle("PCs 1 & 2")+
  theme(plot.title = element_text(hjust = 0.5)) +
  scale_colour_viridis_d() 
  
pc12

```

```{r}
pc12<-ggplot(sub12, aes(x = val_x, y = val_y, color = DI, shape=`library number`)) +
#  geom_point(size = 2, alpha=0.75) +
  geom_text_repel(aes(label=Letter), max.overlaps = Inf, alpha=0.75) +
  scale_fill_discrete(na.value = "white") + 
  theme_bw()+
  theme(panel.grid=element_blank())+
  xlab(paste("PC1", " ", round((100*var[1]),2), "%", sep = ""))+
  ylab(paste("PC2", " ", round((100*var[2]),2), "%", sep = ""))+
  ggtitle("PCs 1 & 2")+
  theme(plot.title = element_text(hjust = 0.5)) +
  scale_colour_viridis_d() +
  facet_wrap(.~DI)
  
pc12

```


### Kin?

```{r}
kin<-npyLoad("outputs/601/test96.kinship.kinship.npy")
excl<-npyLoad("outputs/601/test96-kinremoved.unrelated.npy") %>% as_tibble()

#Convert upper half of matrix to data frame compatible 
kinrind<-which(upper.tri(kin, diag = TRUE), arr.ind = TRUE)
kin<-cbind(kinrind, kin[kinrind])
kindf<-as_tibble(kin) %>%
  rename(Ind1=row, Ind2=col, Kinship=V3) %>%
  filter(Ind1 != Ind2) %>%
  mutate(Mean=mean(Kinship))

#Can I replace Ind1 and Ind2 with individual labels???
df<-kindf %>% left_join(pcseqs, by=c("Ind1"="Individual")) %>% 
  left_join(pcseqs, by=c("Ind2"="Individual")) 

#Ahh, and what are our "related" individuals
df %>% filter(Kinship>0.0625) %>% group_by(DI.x, DI.y) %>% summarize(Count=n(), MeanKinship=mean(Kinship))
```

```{r}
df %>% filter(Kinship>0.0625) %>% select(CrossType.x, Letter.x, Number.x, CrossType.y, Letter.y, Number.y, Kinship)
```

```{r}
cols<-viridis(6)
popdf<-df %>% filter(CrossType.x==CrossType.x) %>% mutate(Cross=CrossType.x) 

ggplot(popdf)+geom_histogram(aes(x=Kinship), bins=50, fill="grey", color="grey") +
  geom_vline(xintercept =.25, col=cols[1]) + #Parent offspring/full sibs
  geom_vline(xintercept =.125, col=cols[2]) + #2nd degree
  geom_vline(xintercept =.0625, col=cols[3]) + # 3rd
  geom_vline(xintercept = .0313, col=cols[4]) +
  geom_vline(xintercept = .0156, col=cols[5]) +
  geom_vline(xintercept = 0.0, col=cols[6]) +
  scale_x_continuous(breaks=c(0,0.0156, 0.0313, 0.0625, 0.125, 0.25),
                     labels=c("0 / Unrelated","5th Deg.","4th Deg.","3rd Deg.", "2nd Deg.", "1st Deg.")) +
  theme_bw() +
  theme(axis.text.x = element_text(hjust=1, angle=45)) +
  theme(panel.grid=element_blank()) +
  ylab("Count") +
  xlab("Kinship Coefficient") +
  facet_wrap(Cross ~ ., ncol=2, scales="free_y")

ggsave("outputs/600/kinship.jpg")
```



```{r}
cov<-read_delim("outputs/601/test96-kinremoved.cov", col_names=FALSE, delim=" ") %>% as.matrix()
```


```{r}
meta<-pcseqs
kept<-meta[excl$value!=0,]

write_tsv(kept, "meta/kept96.tsv")

dropped<-meta[excl$value==0,]

#These are the ones that were dropped:
dropped
```

```{r}
pca <- covar2pcs(kept$Individual, cov)

pca_long <- pca$PCs %>%
  tidyr::gather(., key = "PC", "val", -sample)

# then expand a grid of the possible comparisons (ordered)
expg <- expand.grid(sample = pca$PCs$sample,
                    PCx = sprintf("PC-%02d", 1:6),
                    PCy = sprintf("PC-%02d", 1:6),
                    stringsAsFactors = FALSE) %>%
  tibble::as_tibble()

# then left join the pca results onto that
pca_pairs <- dplyr::left_join(expg, pca_long, by = c("sample", "PCx" = "PC")) %>%
  dplyr::rename(val_x = val) %>%
  dplyr::left_join(pca_long, by = c("sample", "PCy" = "PC")) %>%
  dplyr::rename(val_y = val)

pp_meta <- pca_pairs %>%   # just keep the first 6 PCs around
  left_join(., kept, by = c("sample" = "Individual")) 

npc <- 3
pp_meta2 <- pp_meta %>%
  filter( (PCx %in% sprintf("PC-%02d", 1:npc)) & 
            (PCy %in% sprintf("PC-%02d", 1:npc)) )

eig <- eigen(cov, symm = TRUE)
var<-eig$values/sum(eig$values)
cumvar<-cumsum(eig$values)/sum(eig$values)


ggplot(pp_meta2) +
  geom_point(aes(x=val_x, y=val_y, shape=`library number`, color=DI)) +
  coord_fixed() +
  facet_grid (PCx ~ PCy)

```
    
    
```{r}
sub12<-pp_meta2 %>% filter( (PCx =="PC-01") & (PCy =="PC-02") )


pc12<-ggplot(sub12 %>% filter(Coverage < 30), aes(x = val_x, y = val_y, color = Coverage, shape=DI)) +
  geom_point(size = 2, alpha=0.75) +
  scale_fill_discrete(na.value = "white") + 
  theme_bw()+
  theme(panel.grid=element_blank())+
  xlab(paste("PC1", " ", round((100*var[1]),2), "%", sep = ""))+
  ylab(paste("PC2", " ", round((100*var[2]),2), "%", sep = ""))+
  ggtitle("PCs 1 & 2")+
  theme(plot.title = element_text(hjust = 0.5)) +
  scale_colour_viridis_c() +
  facet_wrap(.~`library number`)

pc12
```
