---
title: "703-gwas-redo"
author: "Mac Campbell"
date: "7/14/2022"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, warning=FALSE, message=FALSE)
```

```{r}
library(tidyverse)
library(RColorBrewer)
```

## 91 fish organized like so:

```{r}
kept96<-read_tsv("meta/kept96.tsv")

kept96 %>% group_by(CrossType) %>% summarize(Count=n())
```

91 samples split between HDI and LDI.
"bamlists/test91.bamlist")

Can use 301-do-asso.sh, but need a phenotype file.

0 for controls, 1 for cases      

```{r}
phenos<-kept96 %>% select(`Sample name`, Individual, String, DI) %>% mutate(Pheno=ifelse(DI=="LowDI",0,1))
phenos

write_tsv(phenos %>% select(Pheno), col_names = FALSE, "meta/test91.pheno")
```


Now to run GWAS in ~/dsm-omics/outputs/703 on farm     

```{sh, eval=FALSE}
bash $HOME/dsm-omics/301-do-asso.sh $HOME/dsm-omics/bamlists/test91.bamlist  $HOME/dsm-omics/meta/1mbseqs.txt $HOME/dsm-omics/meta/test91.pheno 
```


### A couple chroms

```{r}
files = list.files(path="outputs/703/", pattern="*.lrt0.gz", full.names = TRUE)
list = lapply(files, read_tsv)
data<-bind_rows(list)
```
```{r}
df <- data %>% mutate(log10p = -log10(dchisq(LRT, df = 1))) %>%
  filter(log10p>=0 & log10p != "Inf") %>%
  mutate(p = dchisq(LRT, df=1)) %>%
  mutate(bonfer = p.adjust(p, method = "bonferroni")) %>% 
  mutate(BHfdr = p.adjust(p, method = "BH")) %>%
  group_by(Chromosome) %>%
  mutate(adjpchrom = p.adjust(p, method = "bonferroni")) %>%
  mutate(BHfdrchrom = p.adjust(p, method = "BH"))

df %>% arrange(-LRT) %>% head(n=20)
```


```{r}
df %>% filter(log10p>20)
```

```{r}
#chroms<-df %>% select(Chromosome, Index) %>% group_by(Chromosome)  %>% summarize(Position=mean(Index))
#chroms
```
colors      
```{r}
nb.cols <- length(unique(df$Chromosome))
mycolors <- colorRampPalette(brewer.pal(8, "Set1"))(nb.cols)
#randomize
mycolors<-sample(mycolors)
```

```{r}
ggplot(df) +
  geom_point(aes(x=Position, y=log10p, color=Chromosome), alpha=0.75, cex=0.5) +
 # geom_hline(yintercept = -log10(0.05/nrow(dddf)), col="black", linetype=2, alpha=0.5) +
#  geom_hline(yintercept= 8, col="black", linetype=1, alpha=0.5) +
  theme_bw() +
#  scale_x_continuous(breaks=chroms$Position, labels=chroms$Chromosome) +
  theme(axis.text.x = element_text(angle=45, hjust = 1, face="bold", size=8)) +
  theme(legend.title = element_blank()) +
  theme(panel.grid = element_blank()) +
  theme(legend.position = "none") +
  scale_color_manual(values=mycolors) +
  ylab("-log10(p)\n") +
  xlab("\nChromosome") +
  ggtitle("HDI vs LDI") +
  theme(plot.title = element_text(hjust=0.5) ) +
  theme(plot.subtitle = element_text(hjust=0.5)) +
  facet_wrap(.~Chromosome, ncol=5) +
  theme(axis.ticks.x = element_blank(), axis.text.x = element_blank())

ggsave("outputs/703/hdi-ldi-91-GWAS.jpeg", width=18, height=12)
```

```{r}
df
```

### Should some other type of GWAS be considered?

I think we can use the covariance matrix as covariates as there is some sort of pattern to it. -cov?   
The covariance matrix to use is outputs/601/test96-kinremoved.cov, see 600 for exporting of PCs

in 704:
```{sh, eval=FALSE}
bash $HOME/dsm-omics/703.1-gwas-with-cov.sh $HOME/dsm-omics/bamlists/test91.bamlist  $HOME/dsm-omics/meta/1mbseqs.txt $HOME/dsm-omics/meta/test91.pheno $HOME/dsm-omics/meta/test91-pcs.tsv
```

This doesn't appear to have any change.
"If you have population structure then you will have to adjust for it in the regression model (-doAssso 2, 4, 5 or 6)"

in 705:
```{sh, eval=FALSE}
bash $HOME/dsm-omics/703.2-doAsso2-withCov.sh $HOME/dsm-omics/bamlists/test91.bamlist  $HOME/dsm-omics/meta/NW.txt $HOME/dsm-omics/meta/test91.pheno $HOME/dsm-omics/meta/test91-pcs.tsv
gunzip -c NW_025814038.1-asso.lrt0.gz | sort -k 7 -n | tail
```
NW_025814038.1	988254	C	A	0.209129	91	10.268268	56/32/3
NW_025814038.1	995938	G	T	0.191798	91	10.392671	56/33/0
NW_025814038.1	150139	A	G	0.327012	91	10.904456	43/36/10
NW_025814038.1	560896	T	C	0.337711	91	11.725648	40/39/8
NW_025814038.1	926467	T	G	0.374247	91	12.018893	33/48/10
NW_025814038.1	797165	G	A	0.355692	91	12.504225	34/48/8
NW_025814038.1	679325	C	T	0.175040	91	12.620963	59/31/0
NW_025814038.1	679705	C	G	0.200590	91	12.636778	59/26/4
NW_025814038.1	564770	C	A	0.335899	91	13.280115	38/45/8
NW_025814038.1	608105	T	C	0.091301	91	14.028644	61/12/0

in 706 (no covariate):     
```{sh, eval=FALSE}
bash $HOME/dsm-omics/703.3-doAsso2.sh $HOME/dsm-omics/bamlists/test91.bamlist  $HOME/dsm-omics/meta/NW.txt $HOME/dsm-omics/meta/test91.pheno
gunzip -c NW_025814038.1-asso.lrt0.gz | sort -k 7 -n | tail
```
NW_025814038.1	996139	C	T	0.193330	91	28.011694	56/33/1
NW_025814038.1	679705	C	G	0.200590	91	28.069243	59/26/4
NW_025814038.1	995393	C	A	0.192474	91	28.181779	57/33/1
NW_025814038.1	990742	T	A	0.193996	91	28.425642	57/32/1
NW_025814038.1	679325	C	T	0.175040	91	28.806605	59/31/0
NW_025814038.1	995938	G	T	0.191798	91	28.994266	56/33/0
NW_025814038.1	988254	C	A	0.209129	91	30.029809	56/32/3
NW_025814038.1	149817	A	G	0.499751	91	30.419134	29/33/28
NW_025814038.1	150139	A	G	0.327012	91	31.711962	43/36/10
NW_025814038.1	233591	T	C	0.425769	91	33.447693	32/30/22

Note: (base) maccamp@farm:~/dsm-omics/outputs/706$ sq | grep "launch failed" | awk '{print $1}' | while read line; do scontrol release $line; done;

Huge, impact of covariates, then in 705:
```{sh, eval=FALSE}
bash $HOME/dsm-omics/703.2-doAsso2-withCov.sh $HOME/dsm-omics/bamlists/test91.bamlist  $HOME/dsm-omics/meta/1mbseqs.txt  $HOME/dsm-omics/meta/test91.pheno $HOME/dsm-omics/meta/test91-pcs.tsv
```


Show change in allele freqs, gwas scores.

```{r}
chroms<-read_tsv("meta/nctochrom.txt")
```

```{r}
files2 = list.files(path="outputs/705/", pattern="*.lrt0.gz", full.names = TRUE)
list2 = lapply(files2, read_tsv)
data2<-bind_rows(list2) %>% mutate(log10p = -log10(dchisq(LRT, df = 1))) %>%
  filter(log10p>=0 & log10p != "Inf") %>%
  mutate(p = dchisq(LRT, df=1)) %>%
  mutate(bonfer = p.adjust(p, method = "bonferroni")) %>% 
  mutate(BHfdr = p.adjust(p, method = "BH")) %>%
  group_by(Chromosome) %>%
  mutate(adjpchrom = p.adjust(p, method = "bonferroni")) %>%
  mutate(BHfdrchrom = p.adjust(p, method = "BH"))

top2<-data2 %>% arrange(-LRT) %>% head(n=20)
top2
```


```{r}
diffs<-read_csv("outputs/702/diffs.csv") %>% mutate(Diff=MajorFrequencyHDI-MajorFrequency)
```
```{r}
dd<-top2 %>% select(Chromosome, Position, LRT)
dd
```

```{r}
dd %>% left_join(diffs)
```

```
